pkg/process/procutil/gpu_detector.go

```diff
@@ -0,0 +1,78 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package procutil
+
+import (
+	"sync"
+
+	workloadmeta "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type GPUDetector struct {
+	DetectedGPU bool
+	gpuEventsCh chan workloadmeta.EventBundle
+	stopCh      chan struct{}
+	mu          sync.Mutex
+	wmeta       workloadmeta.Component
+}
+
+func NewGPUDetector(wmeta workloadmeta.Component) *GPUDetector {
+	return &GPUDetector{
+		stopCh: make(chan struct{}),
+		wmeta:  wmeta,
+	}
+}
+
+func (g *GPUDetector) Run() {
+	// TODO: ensure this is correct
+	filter := workloadmeta.NewFilterBuilder().
+		SetSource(workloadmeta.SourceRuntime).
+		SetEventType(workloadmeta.EventTypeSet).
+		AddKind(workloadmeta.KindGPU).
+		Build()
+
+	g.gpuEventsCh = g.wmeta.Subscribe(
+		"gpu-detector-set-gpu",
+		workloadmeta.NormalPriority,
+		filter,
+	)
+	for {
+		select {
+		case eventBundle, ok := <-g.gpuEventsCh:
+			if !ok {
+				return
+			}
+			for _, event := range eventBundle.Events {
+				gpu, ok := event.Entity.(*workloadmeta.GPU)
+				if !ok {
+					log.Debugf("Expected workloadmeta.GPU got %T, skipping", event.Entity)
+					continue
+				}
+				g.mu.Lock()
+				// TODO: change into a map storing GPU info
+				g.DetectedGPU = true
```

Comment:
Instead of using a mutex, you can simply use an [atomic](https://pkg.go.dev/sync/atomic). You also need to guard the reads maybe through a getter.

Reply:
I'm going to transition from the boolean to a map (to hold pid : gpu_tags), and check if the map was empty to signify whether a GPU was detected. That way I don't have to get each GPU's information separately later on which is duplicate logic. 

For the map, instead of an `atomic`, I'll shard the map (synchronize based on map key / using sync.Map etc.) to best guard the read/writes.

Reply:
we can just query WMS directly from the process checks. We don't need to store yet another map as this data should be available from WMS.